server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java

```diff
@@ -1378,32 +1381,89 @@ public void remove() {
                 }
             };
         } else {
-            return new Iterator<ShardRouting>() {
-                @Override
-                public boolean hasNext() {
-                    while (!queue.isEmpty()) {
-                        if (queue.peek().hasNext()) {
+            if (shardMovementStrategy == BalancedShardsAllocator.ShardMovementStrategy.REPLICA_FIRST) {
+                return new Iterator<ShardRouting>() {
```

Comment:
This iterator is identical to the iterator built within primary_first other than the queue names & the condition at which elements are inserted.  Can we refactor this to a private builder function that takes a function to determine element insertion?

Something like...
```java
private Iterator<ShardRouting> buildIteratorForMovementStrategy(Predicate<ShardRouting> shouldInsert)
```